test/capi/capi_models.cxx

```diff
@@ -280,6 +279,220 @@ BOOST_AUTO_TEST_CASE(test_boosted_trees_double) {
   }
 }
 
+BOOST_AUTO_TEST_CASE(test_recommender) {
+  for (const char* model_name :
+       {"popularity", "item_similarity"}) {
+
+    tc_error* error = NULL;
+
+    CAPI_CHECK_ERROR(error);
+
+    std::vector<std::pair<std::string, std::vector<int64_t> > > data = {
+        {"user_id", {1, 1, 2, 2, 2, 3}},
+        {"item_id", {1, 2, 2, 3, 4, 3}},
+        {"target", {0, 1, 1, 1, 1, 0}}};
+
+    tc_sframe* sf = make_sframe_integer(data);
+
+    tc_parameters* args = tc_parameters_create_empty(&error);
+
+    CAPI_CHECK_ERROR(error);
+
+    // Add in the sframe; then destroy it when we're done with it.
+    {
```

Comment:
I'm not really understanding why your creating a namespace here (i.e. using the curly braces) or why you're doing it on line 307 and line 312. I just don't think it adds anything and find it's kind of confusing. 

I like your usage on line 318 and 329 were you allocation something at the beginning and deallocate it at the end.

Reply:
[Line 65](https://github.com/apple/turicreate/blob/master/test/capi/capi_models.cxx) of the original capi_models.cxx, for instance, does this:
   
> // Add in the sframe; then destroy it when we're done with it.
>     {
>       tc_parameters_add_sframe(args, "data", sf, &error);
>       CAPI_CHECK_ERROR(error);
>     }

I assumed this format to be by convention, so I was keeping consistent with it. But I'll remove the namespaces for tc_parameters_add_sframe() in lines 302, 307 and 312 as you've mentioned.

Reply:
It's more than just formatting. Adding those curly braces introduces a namespace; things created inside can not be accessed from the outside. This can be very useful for keeping long functions sane.

Reply:
Also note that it controls variable construction/destruction implicitly with scoping - meaning a variable inside this block won't run its constructor until the block runs, and will run its destructor at the end of the block. This can be useful (especially in switch/case statements, where you don't want to initialize variables for cases that aren't currently the one being evaluated). Note that `{ }` always creates a scoping block in C/C++, whether it's for a `for`, `if`, a function, or just by itself (like here), and the semantics are the same. (Python and JavaScript are quite different in this respect.)

Reply:
Thank you so much for explaining @TobyRoseman @znation

Reply:
FYI, to avoid future confusion, the C++ construct we're discussing is just a [C++ block](https://en.cppreference.com/w/cpp/language/scope) (compound statement).

In C++, a [namespace](https://en.cppreference.com/w/cpp/language/namespace) is a specific construct always introduced by the `namespace` keyword.

As Toby and Zach discussed, the primary reason to introduce a block (that is not otherwise required by C++ syntax) is to control the lifetime of variables declared inside.

Reply:
That's completely correct: I should have said "scope" not "namespace".